### PR TITLE
Adding the new feature of KFUZZ

### PR DESCRIPTION
**Pull request by KFuzzing**
Hi @vanhauser-thc,

I propose a new fuzzer called KFUZZ, which has achieved promising results on FuzzBench. I am wondering if it's possible to integrate it into AFL++ as one of the modes? I have already implemented KFUZZ on the latest version of AFL++. KFUZZ can be easily integrated into the existing AFL++ as a mode. Additionally, KFUZZ achieves the best results when used in conjunction with cmplog.

## Introduction
KFUZZ proposes a new seed-saving strategy based on AFL++. In addition to saving mutated inputs that discover new paths or edges into fuzzing queue, we also saving certain high-potential mutated inputs, even if they do not find new paths or edges. The potential of these inputs is determined by their execution paths and the characteristics of their original seeds.

To use this new seed-saving strategy, simply add the `-k` flag when running afl-fuzz. For optimal performance, we recommend using this strategy in combination with `cmplog` mode:
```
  ./afl-fuzz -i seeds_dir -o output_dir -k -c /path/to/program_cmp -- \
  /path/to/tested/program [...program's cmdline...]
```

## Evaluation on FuzzBench
To validate the effectiveness of KFUZZ, we conducted experiments on FuzzBench (Report: https://www.fuzzbench.com/reports/experimental/2024-10-14-kfuzz-1/index.html). For fairness, since the version of AFL++ on FuzzBench is 56d5aa3101945e81519a3fac8783d0d8fad82779, the version of KFUZZ used in the experiment was also based on this commit. KFUZZ achieved the highest avg. score (95.67) as well as the highest avg. rank (2.1) on FuzzBench. We have summarized the experimental results of KFUZZ and AFL++ in terms of "unique coverage" and "who achieves higher coverage", with the results presented in the table below.

The prerequisite for using KFUZZ is having at least two initial seeds. In FuzzBench, 15 out of 21 target programs meet this condition. Among these 15 programs, KFUZZ discovered more coverage than AFL++ in 11 of them. Additionally, KFUZZ found more unique coverage than AFL++ in 13 programs. KFUZZ discovered a total of 919 branches that AFL++ could not cover, while AFL++ found only 384 branches that KFUZZ could not cover.

| Programs | KFUZZ-AFL++ [^1] | AFL++-KFUZZ [^2] | Who achieves higher coverage [^3] |
| -------- | -------- | -------- | -------- |
| bloaty_fuzz_target | **31** | 0 | KFUZZ |
| curl_curl_fuzzer_http | **115** | 9 | KFUZZ |
| freetype2_ftfuzzer | **130** | 119 | KFUZZ |
| harfbuzz_hb-shape-fuzzer  | **9** | 7 | AFL++ |
| turbo_libjpeg_turbo_fuzzer  | 0 | 0 | AFL++ |
| libpcap_fuzz_both | **142** | 7 | KFUZZ |
| libpng_libpng_read_fuzzer  | **1** | 0 | KFUZZ |
| libxml2_xml | **23** | 14 | KFUZZ |
| libxslt_xpath | 6 | **18** | AFL++ |
| openssl_x509 | 0 | 0 | KFUZZ |
| openthread_ot | **313** | 1 | KFUZZ |
| proj4_proj_crs_to_crs_fuzzer  | 99 | **173** | KFUZZ |
| sqlite3_ossfuzz | **50** | 29 | KFUZZ |
| systemd_fuzz | 0 | 0 | KFUZZ |
| woff2_convert_woff2ttf_fuzzer  | 0 | **7** | AFL++ |
| Total unique coverage | **919** | 384 |  |

[^1]: The number of code branches covered by KFUZZ but not by AFL++.
[^2]: The number of code branches covered by AFL++ but not by KFUZZ.
[^3]: Who discovered higher coverage, KFUZZ or AFL++?

### CODE REVIEW

File: src/afl-fuzz-state.c [hunk 1]
```diff
@@ -148,6 +148,12 @@ void afl_state_init(afl_state_t *afl, uint32_t map_size) {
 
   list_append(&afl_states, afl);
 
+  afl->new_edges_found_idx = 0;
+  afl->totalOtherFnameNum = 0;
+  afl->normal_mode = 0; // whether we should enter normal fuzzing
+
+  afl->k_mode = 1; // default setting: Using k_mode
```
Comment (vanhauser-thc) [inline src/afl-fuzz-state.c]: you are making this the default, making the -k parameter obsolete and preventing to disable k mode

File: src/afl-fuzz.c [hunk 2]
```diff
@@ -2410,6 +2414,8 @@ int main(int argc, char **argv_orig, char **envp) {
 
   afl->start_time = get_cur_time();
 
+  srand(time(NULL));
```
Comment (vanhauser-thc) [inline src/afl-fuzz.c:2421]: you can use our rng mechanism, it is faster than rand

**Comment by vanhauser-thc** [comment]
thanks for the PR, I changed the base to dev because we only accept PRs there.
Looks good IMHO, I will do another fuzzbench run though just to be sure.

**Comment by vanhauser-thc** [comment]
btw I started it here, let's wait until after the weekend so the results are final:
https://www.fuzzbench.com/reports/experimental/2024-10-17-aflpp2/index.html
(newer afl++ commit id, I hope this does not have a bigger impact. otherwise we do another round and you your repo to current)

File: src/afl-fuzz-state.c (hunk 1 above)
Comment (KFuzzing) [reply src/afl-fuzz-state.c]: I'm sorry I forgot to comment out this line.

**Comment by KFuzzing** [comment]
Thank you for your reply. I hope I can make some contributions to AFL++. I am also continuing to refine the strategy for KFUZZ.

**Comment by KFuzzing** [comment]
Based on the current experimental results, different versions of AFL++ show noticeable performance differences on some programs. Since the initial seed count for `lcms_cms_transform_fuzzer` is less than 2, KFUZZ's strategy is not triggered (KFUZZ is equivalent to the old version of AFL++). However, they exhibit significantly inconsistent coverage performance: the median coverage for the new version of AFL++ is 1900, while the old version of AFL++ is only 1500.

**Comment by vanhauser-thc** [comment]
From the current results I think it makes sense if you update kfuzz to the current AFL++ dev code state as I made some improvements to seed selection some time ago

**Comment by KFuzzing** [comment]
I think so too. I will finish it as soon as possible.

**Comment by KFuzzing** [comment]
I have updated KFUZZ to the current AFL++ dev code state and added some changes (new flag `-H`). Maybe we can run two versions of KFUZZ: (1) with `-k`; (2) with `-k -H`. Thus,  I can more easily quantify the importance of each part.

Note that I have disabled `afl->k_mode = 1;`.

**Comment by vanhauser-thc** [comment]
I am not on my phone until Monday :)
I started a run with the new commit id. The winner will get another run against kfuzz -H.

please check the ci errors

**Comment by KFuzzing** [comment]
Hi @vanhauser-thc,

I found the you use the wrong link of KFUZZ in `builder.Dockerfile`. The right setting is `git clone https://github.com/KFuzzing/AFLplusplus /afl` instead of `git clone https://github.com/KFuzzing/KFuzz /afl`.  `KFuzzing/KFuzz` is based on the 56d5aa3101945e81519a3fac8783d0d8fad82779 of AFL++. And the `KFuzzing/AFLplusplus` is based on the `dev` branch of AFL++.

The right setting should be: 
```
RUN git clone https://github.com/KFuzzing/AFLplusplus /afl && \
    cd /afl && \
    git checkout 9c57ed88fb9daaf163f0a799439fe8aaf9796b71 || \
    true
```

**Comment by KFuzzing** [comment]
In the attachment, I provided the settings for the fuzzers in FuzzBench, which you can refer to. All the fuzzers are based on AFL++'s a11488b9dcc71e5b52876e8c11ea6ee231e433ba.
[fuzzers.zip](https://github.com/user-attachments/files/17459234/fuzzers.zip)

**Comment by KFuzzing** [comment]
It looks like the experiments have run successfully based on the log in [2024-10-21-aflpp](https://storage.googleapis.com/fuzzbench-data/index.html?prefix=2024-10-21-aflpp/)  , but I can not find the corresponding report. Could you please have a look?

Thank you!

**Comment by vanhauser-thc** [comment]
that is a fuzzbench issue I have seen before. once the run is finished it will pop up, 1-2 more days.

**Comment by KFuzzing** [comment]
> that is a fuzzbench issue I have seen before. once the run is finished it will pop up, 1-2 more days.

Thanks!

**Comment by vanhauser-thc** [comment]
the report is up but no kfuzz fuzzer results are displayes :-( ... I am downloading the logs and checking what the issue is.

**Comment by vanhauser-thc** [comment]
Here are the raw coverage results (only completed runs):
```
bloaty_fuzz_target.TXT
=======================
aflplusplus2 8.33%
aflplusplus2 8.20%
aflplusplus2 8.26%
aflplusplus2 8.11%
aflplusplus2 8.31%
aflplusplus2 8.25%
aflplusplus2 8.29%
aflplusplus2 8.22%
aflplusplus2 8.11%
kfuzz1 8.21%
kfuzz1 8.17%
kfuzz1 8.21%
kfuzz1 8.22%
kfuzz1 8.22%
kfuzz1 8.17%
kfuzz1 8.24%
kfuzz2 8.17%
kfuzz2 8.19%
kfuzz2 8.20%
kfuzz2 8.25%
kfuzz2 8.13%
kfuzz2 8.27%
kfuzz2 8.25%
kfuzz2 8.24%

curl_curl_fuzzer_http.TXT
=======================
aflplusplus2 11.17%
aflplusplus2 11.30%
aflplusplus2 11.17%
aflplusplus2 11.23%
aflplusplus2 11.31%
aflplusplus2 11.30%
aflplusplus2 11.19%
aflplusplus2 11.08%
kfuzz1 11.04%
kfuzz1 11.05%
kfuzz1 10.85%
kfuzz1 11.15%
kfuzz1 11.08%
kfuzz2 11.25%
kfuzz2 10.97%
kfuzz2 11.24%
kfuzz2 11.03%
kfuzz2 11.29%
kfuzz2 11.25%
kfuzz2 11.26%
kfuzz2 11.13%
kfuzz2 11.20%

freetype2_ftfuzzer.TXT
=======================
aflplusplus2 38.64%
aflplusplus2 39.42%
aflplusplus2 36.50%
aflplusplus2 36.64%
aflplusplus2 38.26%
aflplusplus2 38.59%
aflplusplus2 37.24%
aflplusplus2 38.51%
kfuzz1 35.36%
kfuzz1 33.89%
kfuzz1 34.08%
kfuzz1 34.82%
kfuzz1 35.66%
kfuzz1 33.96%
kfuzz1 35.06%
kfuzz1 34.54%
kfuzz1 34.26%
kfuzz2 38.17%
kfuzz2 38.04%
kfuzz2 37.85%
kfuzz2 37.63%
kfuzz2 38.16%

harfbuzz_hb-shape-fuzzer.TXT
=======================
bitmap_cvg        : 54.48%
kfuzz1 54.52%
kfuzz1 54.31%
kfuzz1 54.35%
kfuzz1 54.12%
kfuzz1 54.63%
kfuzz1 54.70%
kfuzz1 54.62%
kfuzz2 54.45%
kfuzz2 54.23%
kfuzz2 54.71%
kfuzz2 54.25%
kfuzz2 54.66%
kfuzz2 53.92%
kfuzz2 54.47%
kfuzz2 54.68%
kfuzz2 54.11%
kfuzz2 54.82%

jsoncpp_jsoncpp_fuzzer.TXT
=======================
aflplusplus2 21.93%
aflplusplus2 21.93%
aflplusplus2 21.93%
aflplusplus2 21.93%
aflplusplus2 21.93%
aflplusplus2 21.93%
aflplusplus2 21.93%
aflplusplus2 21.93%
aflplusplus2 21.93%
aflplusplus2 21.93%
aflplusplus2 21.93%
kfuzz1 21.86%
kfuzz1 21.86%
kfuzz1 21.70%
kfuzz1 21.79%
kfuzz1 21.79%
kfuzz1 21.70%
kfuzz1 21.79%
kfuzz2 21.93%
kfuzz2 21.93%
kfuzz2 21.93%
kfuzz2 21.93%
kfuzz2 21.93%
kfuzz2 21.93%
kfuzz2 21.93%
kfuzz2 21.93%
kfuzz2 21.93%
kfuzz2 21.93%
kfuzz2 21.93%

lcms_cms_transform_fuzzer.TXT
=======================
aflplusplus2 17.38%
aflplusplus2 22.93%
aflplusplus2 16.77%
aflplusplus2 22.56%
aflplusplus2 21.00%
aflplusplus2 21.86%
aflplusplus2 22.25%
aflplusplus2 22.02%
aflplusplus2 21.77%
kfuzz1 15.11%
kfuzz1 14.52%
kfuzz1 10.25%
kfuzz1 9.80%
kfuzz1 9.91%
kfuzz1 16.00%
kfuzz1 16.87%
kfuzz1 10.43%
kfuzz1 14.83%
kfuzz1 16.14%
kfuzz1 15.55%
kfuzz2 22.88%
kfuzz2 16.62%
kfuzz2 22.35%
kfuzz2 16.38%
kfuzz2 22.12%
kfuzz2 17.38%
kfuzz2 22.89%
kfuzz2 22.72%
kfuzz2 22.52%
kfuzz2 22.79%

libjpeg-turbo_libjpeg_turbo_fuzzer.TXT
=======================
aflplusplus2 17.57%
aflplusplus2 17.57%
aflplusplus2 17.57%
aflplusplus2 17.57%
aflplusplus2 17.57%
aflplusplus2 17.56%
aflplusplus2 17.56%
aflplusplus2 17.57%
aflplusplus2 17.57%
aflplusplus2 17.57%
aflplusplus2 17.57%
kfuzz1 17.57%
kfuzz1 17.57%
kfuzz1 17.57%
kfuzz1 17.57%
kfuzz1 17.61%
kfuzz1 17.51%
kfuzz1 17.57%
kfuzz1 17.67%
kfuzz2 17.57%
kfuzz2 17.57%
kfuzz2 17.57%
kfuzz2 17.57%
kfuzz2 17.57%
kfuzz2 17.54%
kfuzz2 17.56%
kfuzz2 17.57%
kfuzz2 17.57%

libpcap_fuzz_both.TXT
=======================
aflplusplus2 28.73%
aflplusplus2 29.02%
aflplusplus2 28.24%
aflplusplus2 26.54%
aflplusplus2 29.72%
aflplusplus2 27.08%
aflplusplus2 28.93%
aflplusplus2 28.33%
aflplusplus2 27.93%
aflplusplus2 26.36%
kfuzz1 26.81%
kfuzz1 27.95%
kfuzz1 25.25%
kfuzz1 27.33%
kfuzz1 29.30%
kfuzz2 29.21%
kfuzz2 28.53%
kfuzz2 27.90%
kfuzz2 27.27%
kfuzz2 26.56%
kfuzz2 27.31%
kfuzz2 28.61%

libpng_libpng_read_fuzzer.TXT
=======================
aflplusplus2 29.37%
aflplusplus2 29.39%
aflplusplus2 29.36%
aflplusplus2 29.36%
aflplusplus2 29.34%
aflplusplus2 29.32%
aflplusplus2 29.32%
aflplusplus2 29.52%
aflplusplus2 29.36%
kfuzz1 29.13%
kfuzz1 29.24%
kfuzz1 29.29%
kfuzz1 29.26%
kfuzz1 29.29%
kfuzz1 29.29%
kfuzz1 29.27%
kfuzz1 29.31%
kfuzz1 29.31%
kfuzz1 29.32%
kfuzz1 29.26%
kfuzz1 29.26%
kfuzz2 29.37%
kfuzz2 29.34%
kfuzz2 29.36%
kfuzz2 29.41%
kfuzz2 30.82%
kfuzz2 29.32%
kfuzz2 29.39%
kfuzz2 29.34%
kfuzz2 29.31%

libxml2_xml.TXT
=======================
aflplusplus2 19.84%
aflplusplus2 19.75%
aflplusplus2 19.83%
aflplusplus2 19.62%
aflplusplus2 19.79%
aflplusplus2 19.91%
aflplusplus2 19.94%
aflplusplus2 20.00%
kfuzz1 19.76%
kfuzz1 19.80%
kfuzz1 19.67%
kfuzz1 19.87%
kfuzz1 19.81%
kfuzz1 19.97%
kfuzz1 19.76%
kfuzz1 19.82%
kfuzz2 19.76%
kfuzz2 19.81%
kfuzz2 19.71%
kfuzz2 19.93%
kfuzz2 19.76%
kfuzz2 19.78%
kfuzz2 19.81%
kfuzz2 19.82%
kfuzz2 19.93%

libxslt_xpath.TXT
=======================
aflplusplus2 17.67%
aflplusplus2 17.72%
aflplusplus2 17.67%
aflplusplus2 17.93%
aflplusplus2 17.68%
aflplusplus2 17.69%
aflplusplus2 17.88%
aflplusplus2 17.82%
kfuzz1 17.76%
kfuzz1 17.58%
kfuzz1 17.42%
kfuzz1 17.34%
kfuzz1 17.47%
kfuzz1 17.14%
kfuzz1 17.39%
kfuzz2 17.73%
kfuzz2 17.66%
kfuzz2 17.67%
kfuzz2 17.59%
kfuzz2 17.72%

mbedtls_fuzz_dtlsclient.TXT
=======================
aflplusplus2 11.84%
aflplusplus2 11.98%
aflplusplus2 11.73%
aflplusplus2 11.83%
aflplusplus2 12.01%
kfuzz1 11.65%
kfuzz1 11.76%
kfuzz1 11.81%
kfuzz1 11.81%
kfuzz2 11.78%
kfuzz2 11.94%
kfuzz2 12.11%
kfuzz2 12.15%
kfuzz2 11.90%
kfuzz2 11.84%
kfuzz2 11.88%

openssl_x509.TXT
=======================
aflplusplus2 8.67%
aflplusplus2 8.66%
aflplusplus2 8.70%
aflplusplus2 8.68%
kfuzz1 8.70%
kfuzz1 8.68%
kfuzz1 8.68%
kfuzz1 8.70%
kfuzz1 8.67%
kfuzz1 8.67%
kfuzz1 8.69%
kfuzz1 8.70%
kfuzz1 8.67%
kfuzz2 8.70%
kfuzz2 8.68%
kfuzz2 8.68%
kfuzz2 8.68%
kfuzz2 8.71%

openthread_ot-ip6-send-fuzzer.TXT
=======================
aflplusplus2 15.27%
aflplusplus2 15.42%
aflplusplus2 15.22%
aflplusplus2 16.67%
aflplusplus2 15.36%
aflplusplus2 16.88%
aflplusplus2 15.39%
aflplusplus2 15.21%
aflplusplus2 15.21%
aflplusplus2 16.75%
aflplusplus2 15.34%
kfuzz1 14.36%
kfuzz1 14.38%
kfuzz1 14.36%
kfuzz1 14.36%
kfuzz1 14.38%
kfuzz1 14.37%
kfuzz1 14.38%
kfuzz2 15.14%
kfuzz2 17.04%
kfuzz2 15.33%
kfuzz2 15.20%
kfuzz2 15.20%

proj4_proj_crs_to_crs_fuzzer.TXT
=======================
aflplusplus2 16.54%
aflplusplus2 16.26%
aflplusplus2 15.71%
aflplusplus2 15.85%
aflplusplus2 15.74%
aflplusplus2 16.10%
aflplusplus2 16.00%
aflplusplus2 15.98%
aflplusplus2 16.52%
aflplusplus2 16.18%
aflplusplus2 16.49%
aflplusplus2 15.76%
aflplusplus2 16.21%
kfuzz1 13.66%
kfuzz1 13.82%
kfuzz1 14.06%
kfuzz1 14.00%
kfuzz1 13.74%
kfuzz1 13.77%
kfuzz1 14.13%
kfuzz2 16.49%
kfuzz2 16.46%
kfuzz2 15.86%
kfuzz2 16.45%
kfuzz2 15.51%
kfuzz2 15.89%
kfuzz2 16.59%
kfuzz2 16.52%
kfuzz2 15.83%
kfuzz2 16.43%
kfuzz2 16.24%
kfuzz2 16.02%
kfuzz2 16.22%

re2_fuzzer.TXT
=======================
aflplusplus2 49.20%
aflplusplus2 49.51%
aflplusplus2 49.50%
aflplusplus2 49.51%
aflplusplus2 49.55%
aflplusplus2 49.46%
aflplusplus2 49.54%
aflplusplus2 49.45%
aflplusplus2 49.37%
aflplusplus2 49.48%
kfuzz1 49.05%
kfuzz1 49.39%
kfuzz1 49.37%
kfuzz1 49.37%
kfuzz1 49.30%
kfuzz1 49.47%
kfuzz1 49.32%
kfuzz2 49.48%
kfuzz2 49.38%
kfuzz2 49.53%
kfuzz2 49.56%
kfuzz2 49.51%
kfuzz2 49.58%
kfuzz2 49.52%

sqlite3_ossfuzz.TXT
=======================
aflplusplus2 51.74%
aflplusplus2 50.68%
aflplusplus2 50.70%
aflplusplus2 51.07%
aflplusplus2 50.92%
aflplusplus2 51.24%
aflplusplus2 50.43%
aflplusplus2 49.25%
aflplusplus2 51.17%
aflplusplus2 51.44%
aflplusplus2 50.54%
aflplusplus2 51.36%
aflplusplus2 52.03%
aflplusplus2 49.86%
kfuzz1 50.54%
kfuzz1 50.57%
kfuzz1 51.05%
kfuzz1 50.65%
kfuzz1 50.04%
kfuzz1 51.22%
kfuzz1 51.06%
kfuzz2 50.34%
kfuzz2 50.28%
kfuzz2 51.59%
kfuzz2 51.59%
kfuzz2 51.40%
kfuzz2 51.85%
kfuzz2 50.80%

systemd_fuzz-link-parser.TXT
=======================
aflplusplus2 2.57%
aflplusplus2 2.55%
aflplusplus2 2.53%
kfuzz1 2.48%
kfuzz1 2.51%
kfuzz1 2.51%
kfuzz1 2.51%
kfuzz1 2.52%
kfuzz1 2.52%
kfuzz1 2.50%
kfuzz2 2.52%
kfuzz2 2.53%
kfuzz2 2.54%
kfuzz2 2.50%
kfuzz2 2.48%
kfuzz2 2.55%
kfuzz2 2.54%
kfuzz2 2.56%
kfuzz2 2.50%
kfuzz2 2.53%
kfuzz2 2.55%

vorbis_decode_fuzzer.TXT
=======================
aflplusplus2 13.59%
aflplusplus2 13.44%
aflplusplus2 13.47%
aflplusplus2 13.55%
aflplusplus2 13.43%
aflplusplus2 13.47%
aflplusplus2 13.45%
aflplusplus2 13.45%
aflplusplus2 13.46%
aflplusplus2 13.47%
aflplusplus2 13.47%
aflplusplus2 13.40%
kfuzz1 13.40%
kfuzz1 13.43%
kfuzz1 13.48%
kfuzz1 13.48%
kfuzz1 13.47%
kfuzz1 13.46%
kfuzz1 13.55%
kfuzz1 13.50%
kfuzz1 13.46%
kfuzz1 13.46%
kfuzz1 13.51%
kfuzz1 13.44%
kfuzz2 13.43%
kfuzz2 13.43%
kfuzz2 13.55%
kfuzz2 13.48%
kfuzz2 13.57%
kfuzz2 13.46%

woff2_convert_woff2ttf_fuzzer.TXT
=======================
aflplusplus2 20.78%
aflplusplus2 21.89%
aflplusplus2 20.45%
aflplusplus2 20.62%
aflplusplus2 20.46%
aflplusplus2 21.85%
aflplusplus2 20.51%
aflplusplus2 20.52%
aflplusplus2 21.91%
aflplusplus2 20.75%
aflplusplus2 20.67%
aflplusplus2 20.44%
aflplusplus2 20.59%
aflplusplus2 20.54%
kfuzz1 20.74%
kfuzz1 22.66%
kfuzz1 22.73%
kfuzz1 20.73%
kfuzz1 20.84%
kfuzz1 22.21%
kfuzz1 22.71%
kfuzz2 21.73%
kfuzz2 21.95%
kfuzz2 20.46%
kfuzz2 20.49%
kfuzz2 21.09%
kfuzz2 22.01%

zlib_zlib_uncompress_fuzzer.TXT
=======================
aflplusplus2 40.37%
aflplusplus2 40.26%
aflplusplus2 40.77%
kfuzz1 40.87%
kfuzz1 40.57%
kfuzz1 40.97%
kfuzz1 40.67%
kfuzz1 40.37%
kfuzz1 40.77%
kfuzz1 40.37%
kfuzz2 40.87%
kfuzz2 40.77%
kfuzz2 40.87%
kfuzz2 40.67%
kfuzz2 40.87%
kfuzz2 40.67%
kfuzz2 40.67%
kfuzz2 40.37%
```

**Comment by vanhauser-thc** [comment]
a quick look shows that kfuzz2 is better than kfuzz1, however I do not see a noticable difference between kfuzz2 and aflplusplus

**Comment by KFuzzing** [comment]
I'm not sure if there is an issue with my recent code changes. Could you help me run another comparison experiment?

[fuzzers-2.zip](https://github.com/user-attachments/files/17514605/fuzzers-2.zip)

**Comment by vanhauser-thc** [comment]
for whatever reason this is running with a visible intermediate report: https://www.fuzzbench.com/reports/experimental/2024-10-25-aflpp/index.html
should have useful results in about 50 hours.

**Comment by KFuzzing** [comment]
Thank you for your help. The results don't look as expected. I'll look into them.

**Comment by vanhauser-thc** [comment]
Maybe you compared against an afl++ commit on fuzzbench that was buggy or a change you did degraded performance. Don’t give up :)

**Comment by KFuzzing** [comment]
> Maybe you compared against an afl++ commit on fuzzbench that was buggy or a change you did degraded performance. Don’t give up :)

Thank you very much for your encouragement!!! I will keep going and hope to contribute to AFL++ in the future.

**Comment by vanhauser-thc** [comment]
I made a new run with kfuzz2 as currently the google cloud seems to be more stable, and voila it runs with intermediate results:
https://www.fuzzbench.com/reports/experimental/2024-10-27-aflpp/index.html
will take another 48 hours for stable results but so far it looks much better

**Comment by vanhauser-thc** [comment]
The new results are much better for kfuzz now, but they are basically equivalent to afl++

**Pull request closed without merging**